supervisor/process.go

```diff
@@ -0,0 +1,408 @@
+package supervisor
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pkg/errors"
+	"os/exec"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const (
+	DefaultLineBufferSize = 16384
+	DefaultStreamChanSize = 200
+)
+
+type ProcessState uint8
+
+const (
+	Initial ProcessState = iota
+	Starting
+	Running
+	Stopping
+	Finished
+	Interrupted
+	Fatal
+)
+
+func (s ProcessState) String() string {
+	switch s {
+	case Initial:
+		return "initial"
+	case Starting:
+		return "starting"
+	case Running:
+		return "running"
+	case Stopping:
+		return "stopping"
+	case Finished:
+		return "finished"
+	case Interrupted:
+		return "interrupted"
+	case Fatal:
+		return "fatal"
+	default:
+		return "unknown"
+	}
+}
```

Comment:
IMO a custom string type makes more sense here, Go iota's are rarely the best way to define constants.
```suggestion
type ProcessState string

const (
	Initial ProcessState = "intial"
	Starting ProcessState = "starting"
	Running ProcessState = "running"
	Stopping ProcessState = "stopping"
	Finished ProcessState = "finished"
	Interrupted ProcessState = "interrupted"
	Fatal ProcessState = "failed"
)
```